File: src/api/links.ts
// Copyright 2026 Oddbit (https://oddbit.id)
// SPDX-License-Identifier: Apache-2.0

import { createRoute, z } from "@hono/zod-openapi";
import { createApiSubApp } from "./sub-app";
import {
  autoLabelLink,
  createLink,
  disableLink,
  enableLink,
  deleteLink,
  getLink,
  getLinkBySlug,
  listLinks,
  listLinksByOwner,
  updateLink,
  addCustomSlugToLink,
  disableSlug,
  enableSlug,
  removeSlug,
} from "../services/link-management";
import { listBundlesForLink } from "../services/bundle-management";
import { handleLinkQr } from "./qr";
import { handlePublicLinkAnalytics, handlePublicLinkTimeline } from "./analytics";
import { fetchPageTitle } from "../title-fetch";
import { fromServiceResult, json } from "./response";
import { requireScope } from "./scope";
import type { Env, TimelineRange } from "../types";
import {
  AddSlugBodySchema,
  BundleSchema,
  ClickStatsSchema,
  CreateLinkBodySchema,
  ErrorResponseSchema,
  IdParamSchema,
  LinkSchema,
  RangeQuerySchema,
  SlugParamSchema,
  SlugSchema,
  TimelineDataSchema,
  UpdateLinkBodySchema,
  paramHook,
} from "./schemas";
export const linksApp = createApiSubApp();

const errorResponses = {
  400: { description: "Validation error.", content: { "application/json": { schema: ErrorResponseSchema } } },
  401: { description: "Missing or invalid bearer token.", content: { "application/json": { schema: ErrorResponseSchema } } },
  403: { description: "Scope insufficient.", content: { "application/json": { schema: ErrorResponseSchema } } },
  404: { description: "Not found.", content: { "application/json": { schema: ErrorResponseSchema } } },
  409: { description: "Conflict.", content: { "application/json": { schema: ErrorResponseSchema } } },
  500: { description: "Server error.", content: { "application/json": { schema: ErrorResponseSchema } } },
};

// ---- POST / (create link) ----

const createLinkRoute = createRoute({
  method: "post",
  path: "/",
  tags: ["links"],
  summary: "Create a short link",
  middleware: [requireScope("create")] as const,
  request: {
    body: { content: { "application/json": { schema: CreateLinkBodySchema } } },
  },
  responses: {
    200: { description: "Existing link returned (duplicate URL).", content: { "application/json": { schema: LinkSchema } } },
    201: { description: "Created.", content: { "application/json": { schema: LinkSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    409: errorResponses[409],
    500: errorResponses[500],
  },
});

linksApp.openapi(createLinkRoute, async (c) => {
  const body = c.req.valid("json") as { url: string; label?: string; slug_length?: number; custom_slug?: string; expires_at?: number; allow_duplicate?: boolean };
  const via = c.req.header("X-Client") === "sdk" ? "sdk" : "api";
  const result = await createLink(c.env, { ...body, created_via: via, created_by: c.var.auth.identity });
  if (result.ok && result.status === 201 && !body.label) {
    c.executionCtx.waitUntil(autoLabelLink(c.env.DB, result.data.id, result.data.url, fetchPageTitle));
  }
  return fromServiceResult(result) as never;
});

// ---- GET / (list links) ----

const listLinksQuery = z.object({
  owner: z.string().optional().openapi({ description: "Filter to links created by this owner identity." }),
}).merge(RangeQuerySchema);

const listLinksRoute = createRoute({
  method: "get",
  path: "/",
  tags: ["links"],
  summary: "List all short links",
  middleware: [requireScope("read")] as const,
  request: { query: listLinksQuery },
  responses: {
    200: { description: "OK.", content: { "application/json": { schema: z.array(LinkSchema) } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    500: errorResponses[500],
  },
});

linksApp.openapi(listLinksRoute, async (c) => {
  const { owner, range } = c.req.valid("query") as { owner?: string; range?: TimelineRange };
  const opts = range ? { range, withDeltaRange: range } : undefined;
  const result = owner ? await listLinksByOwner(c.env, owner, opts) : await listLinks(c.env, opts);
  return fromServiceResult(result) as never;
});

// ---- GET /:id (get link) ----

const getLinkRoute = createRoute({
  method: "get",
  path: "/{id}",
  tags: ["links"],
  summary: "Get a link with click stats",
  middleware: [requireScope("read")] as const,
  request: { params: IdParamSchema, query: RangeQuerySchema },
  responses: {
    200: { description: "OK.", content: { "application/json": { schema: LinkSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
    500: errorResponses[500],
  },
});

linksApp.openapi(getLinkRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  const { range } = c.req.valid("query") as { range?: TimelineRange };
  return fromServiceResult(await getLink(c.env, id, range ? { range } : undefined)) as never;
}, paramHook);

// ---- PUT /:id (update link) ----

const updateLinkRoute = createRoute({
  method: "put",
  path: "/{id}",
  tags: ["links"],
  summary: "Update a link's URL, label, or expiry",
  middleware: [requireScope("create")] as const,
  request: {
    params: IdParamSchema,
    body: { content: { "application/json": { schema: UpdateLinkBodySchema } } },
  },
  responses: {
    200: { description: "Updated.", content: { "application/json": { schema: LinkSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(updateLinkRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  const body = c.req.valid("json") as { url?: string; label?: string | null; expires_at?: number | null };
  return fromServiceResult(await updateLink(c.env, id, body)) as never;
}, paramHook);

// ---- POST /:id/disable ----

const disableLinkRoute = createRoute({
  method: "post",
  path: "/{id}/disable",
  tags: ["links"],
  summary: "Disable a link",
  middleware: [requireScope("create")] as const,
  request: { params: IdParamSchema },
  responses: {
    200: { description: "Disabled.", content: { "application/json": { schema: LinkSchema } } },
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(disableLinkRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  return fromServiceResult(await disableLink(c.env, id, c.var.auth.identity)) as never;
}, paramHook);

// ---- POST /:id/enable ----

const enableLinkRoute = createRoute({
  method: "post",
  path: "/{id}/enable",
  tags: ["links"],
  summary: "Re-enable a disabled link",
  middleware: [requireScope("create")] as const,
  request: { params: IdParamSchema },
  responses: {
    200: { description: "Enabled.", content: { "application/json": { schema: LinkSchema } } },
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(enableLinkRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  return fromServiceResult(await enableLink(c.env, id, c.var.auth.identity)) as never;
}, paramHook);

// ---- DELETE /:id ----

const deleteLinkRoute = createRoute({
  method: "delete",
  path: "/{id}",
  tags: ["links"],
  summary: "Delete a link permanently",
  middleware: [requireScope("create")] as const,
  request: { params: IdParamSchema },
  responses: {
    200: { description: "Deleted.", content: { "application/json": { schema: z.object({ deleted: z.boolean() }) } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(deleteLinkRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  return fromServiceResult(await deleteLink(c.env, id, c.var.auth.identity)) as never;
}, paramHook);

// ---- POST /:id/slugs (add custom slug) ----

const addSlugRoute = createRoute({
  method: "post",
  path: "/{id}/slugs",
  tags: ["slugs"],
  summary: "Add a custom slug to a link",
  middleware: [requireScope("create")] as const,
  request: {
    params: IdParamSchema,
    body: { content: { "application/json": { schema: AddSlugBodySchema } } },
  },
  responses: {
    201: { description: "Slug added.", content: { "application/json": { schema: SlugSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
    409: errorResponses[409],
  },
});

linksApp.openapi(addSlugRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  const { slug } = c.req.valid("json") as { slug: string };
  return fromServiceResult(await addCustomSlugToLink(c.env, id, { slug })) as never;
}, paramHook);

// ---- Slug-scoped param schema ----

const LinkSlugParamsSchema = IdParamSchema
  .extend({ slug: SlugParamSchema.shape.slug })
  .openapi("LinkSlugParams");

// ---- POST /:id/slugs/:slug/disable ----

const disableSlugRoute = createRoute({
  method: "post",
  path: "/{id}/slugs/{slug}/disable",
  tags: ["slugs"],
  summary: "Disable a specific slug on a link",
  middleware: [requireScope("create")] as const,
  request: { params: LinkSlugParamsSchema },
  responses: {
    200: { description: "Disabled.", content: { "application/json": { schema: SlugSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(disableSlugRoute, async (c) => {
  const { id, slug } = c.req.valid("param") as { id: number; slug: string };
  return fromServiceResult(await disableSlug(c.env, id, slug, c.var.auth.identity)) as never;
}, paramHook);

// ---- POST /:id/slugs/:slug/enable ----

const enableSlugRoute = createRoute({
  method: "post",
  path: "/{id}/slugs/{slug}/enable",
  tags: ["slugs"],
  summary: "Re-enable a disabled slug on a link",
  middleware: [requireScope("create")] as const,
  request: { params: LinkSlugParamsSchema },
  responses: {
    200: { description: "Enabled.", content: { "application/json": { schema: SlugSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(enableSlugRoute, async (c) => {
  const { id, slug } = c.req.valid("param") as { id: number; slug: string };
  return fromServiceResult(await enableSlug(c.env, id, slug, c.var.auth.identity)) as never;
}, paramHook);

// ---- DELETE /:id/slugs/:slug ----

const removeSlugRoute = createRoute({
  method: "delete",
  path: "/{id}/slugs/{slug}",
  tags: ["slugs"],
  summary: "Remove a custom slug from a link",
  middleware: [requireScope("create")] as const,
  request: { params: LinkSlugParamsSchema },
  responses: {
    200: { description: "Removed.", content: { "application/json": { schema: z.object({ removed: z.boolean() }) } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(removeSlugRoute, async (c) => {
  const { id, slug } = c.req.valid("param") as { id: number; slug: string };
  return fromServiceResult(await removeSlug(c.env, id, slug, c.var.auth.identity)) as never;
}, paramHook);

// ---- GET /:id/qr ----

const linkQrRoute = createRoute({
  method: "get",
  path: "/{id}/qr",
  tags: ["qr"],
  summary: "Get a QR code SVG for a link",
  middleware: [requireScope("read")] as const,
  request: {
    params: IdParamSchema,
    query: z.object({
      slug: z.string().regex(/^[a-zA-Z0-9._~-]+$/).optional()
        .openapi({ description: "Optional specific slug. Defaults to the link's primary slug." }),
      size: z.string().regex(/^\d+$/).optional()
        .openapi({ description: "PNG dimensions in pixels (square). Default per server config." }),
    }),
  },
  responses: {
    200: {
      description: "SVG image",
      content: { "image/svg+xml": { schema: z.string() } },
    },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(linkQrRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  return (await handleLinkQr(c.req.raw, c.env, id)) as never;
}, paramHook);

// ---- GET /:id/bundles (list bundles a link belongs to) ----

const listLinkBundlesRoute = createRoute({
  method: "get",
  path: "/{id}/bundles",
  tags: ["links", "bundles"],
  summary: "List bundles a link belongs to",
  middleware: [requireScope("read")] as const,
  request: { params: IdParamSchema },
  responses: {
    200: { description: "OK.", content: { "application/json": { schema: z.array(BundleSchema) } } },
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
  },
});

linksApp.openapi(listLinkBundlesRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  return fromServiceResult(await listBundlesForLink(c.env, id, c.var.auth.identity)) as never;
}, paramHook);

// ---- GET /:id/analytics ----

const linkAnalyticsRoute = createRoute({
  method: "get",
  path: "/{id}/analytics",
  tags: ["links", "analytics"],
  summary: "Get analytics for a link (defaults to all-time)",
  middleware: [requireScope("read")] as const,
  request: { params: IdParamSchema, query: RangeQuerySchema },
  responses: {
    200: { description: "OK", content: { "application/json": { schema: ClickStatsSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
    500: errorResponses[500],
  },
});

linksApp.openapi(linkAnalyticsRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  const { range } = c.req.valid("query") as { range?: string };
  return (await handlePublicLinkAnalytics(c.env, id, range)) as never;
}, paramHook);

// ---- GET /:id/timeline ----

const linkTimelineRoute = createRoute({
  method: "get",
  path: "/{id}/timeline",
  tags: ["links", "analytics"],
  summary: "Get click timeline for a link (defaults to all-time)",
  middleware: [requireScope("read")] as const,
  request: { params: IdParamSchema, query: RangeQuerySchema },
  responses: {
    200: { description: "OK", content: { "application/json": { schema: TimelineDataSchema } } },
    400: errorResponses[400],
    401: errorResponses[401],
    403: errorResponses[403],
    404: errorResponses[404],
    500: errorResponses[500],
  },
});

linksApp.openapi(linkTimelineRoute, async (c) => {
  const { id } = c.req.valid("param") as { id: number };
  const { range } = c.req.valid("query") as { range?: string };
  return (await handlePublicLinkTimeline(c.env, id, range)) as never;
}, paramHook);

// ---- Named exports consumed by admin routes in index.tsx (pending migration in later tasks) ----

export async function handleGetLinkBySlug(env: Env, slug: string): Promise<Response> {
  return fromServiceResult(await getLinkBySlug(env, slug.toLowerCase()));
}

export async function handleListLinks(env: Env, owner?: string): Promise<Response> {
  if (owner) return fromServiceResult(await listLinksByOwner(env, owner));
  return fromServiceResult(await listLinks(env));
}

export async function handleGetLink(env: Env, id: number): Promise<Response> {
  return fromServiceResult(await getLink(env, id));
}

export async function handleCreateLink(request: Request, env: Env, createdVia?: string, createdBy?: string, ctx?: ExecutionContext): Promise<Response> {
  let body: {
    url?: string;
    label?: string;
    slug_length?: number;
    custom_slug?: string;
    expires_at?: number;
    allow_duplicate?: boolean;
  };

  try {
    body = await request.json();
  } catch {
    return json({ error: "Invalid JSON body" }, 400);
  }

  const result = await createLink(env, { ...body, created_via: createdVia, created_by: createdBy });

  if (result.ok && result.status === 201 && !body.label && ctx) {
    ctx.waitUntil(autoLabelLink(env.DB, result.data.id, result.data.url, fetchPageTitle));
  }

  return fromServiceResult(result);
}

export async function handleUpdateLink(request: Request, env: Env, id: number): Promise<Response> {
  let body: { url?: string; label?: string | null; expires_at?: number | null };

  try {
    body = await request.json();
  } catch {
    return json({ error: "Invalid JSON body" }, 400);
  }

  return fromServiceResult(await updateLink(env, id, body));
}

export async function handleDisableLink(env: Env, id: number, identity: string): Promise<Response> {
  return fromServiceResult(await disableLink(env, id, identity));
}

export async function handleEnableLink(env: Env, id: number, identity: string): Promise<Response> {
  return fromServiceResult(await enableLink(env, id, identity));
}

export async function handleDeleteLink(env: Env, id: number, identity: string): Promise<Response> {
  return fromServiceResult(await deleteLink(env, id, identity));
}

File: sdk/python/src/shrtnr/resources/links.py
# Copyright 2026 Oddbit (https://oddbit.id)
# SPDX-License-Identifier: Apache-2.0

"""Links resource: sync and async implementations."""

from __future__ import annotations

# `Links.list` / `AsyncLinks.list` shadow the builtin inside class scope,
# which trips mypy when other methods declare `list[X]` return types. Aliasing
# the builtin here lets those methods keep the PEP 585 `list[X]` style without
# resolving to the method.
from builtins import list as _list
from typing import Any

import httpx

from .._base import (
    _build_auth_headers,
    _build_url,
    parse_json_response,
    parse_text_response,
)
from ..errors import ShrtnrError
from ..models import (
    Bundle,
    ClickStats,
    DeletedResult,
    Link,
    TimelineData,
    TimelineRange,
)


class Links:
    """Synchronous Links resource."""

    def __init__(self, base_url: str, api_key: str, http: httpx.Client) -> None:
        self._base_url = base_url
        self._api_key = api_key
        self._http = http

    def _headers(self) -> dict[str, str]:
        return _build_auth_headers(self._api_key)

    def _json_headers(self) -> dict[str, str]:
        return {**self._headers(), "Content-Type": "application/json"}

    def _url(self, path: str, query: dict[str, str | None] | None = None) -> str:
        return _build_url(self._base_url, path, query)

    def _request(self, method: str, url: str, **kwargs: Any) -> Any:
        try:
            response = self._http.request(method, url, **kwargs)
        except httpx.RequestError as exc:
            raise ShrtnrError(0, str(exc)) from exc
        return parse_json_response(response)

    def _request_text(self, method: str, url: str, **kwargs: Any) -> str:
        try:
            response = self._http.request(method, url, **kwargs)
        except httpx.RequestError as exc:
            raise ShrtnrError(0, str(exc)) from exc
        return parse_text_response(response)

    def get(self, id: int, *, range: TimelineRange | None = None) -> Link:
        """Get a link by ID. Optional range scopes the click-count window."""
        url = self._url(f"/_/api/links/{id}", {"range": range})
        return Link.from_dict(self._request("GET", url, headers=self._headers()))

    def list(self, *, owner: str | None = None, range: TimelineRange | None = None) -> list[Link]:
        """List all links. Filter by owner or click-count range."""
        url = self._url("/_/api/links", {"owner": owner, "range": range})
        data = self._request("GET", url, headers=self._headers())
        return [Link.from_dict(x) for x in (data or [])]

    def create(
        self,
        *,
        url: str,
        label: str | None = None,
        slug_length: int | None = None,
        custom_slug: str | None = None,
        expires_at: int | None = None,
        allow_duplicate: bool | None = None,
    ) -> Link:
        """Create a new short link."""
        body: dict[str, Any] = {"url": url}
        if label is not None:
            body["label"] = label
        if slug_length is not None:
            body["slug_length"] = slug_length
        if custom_slug is not None:
            body["custom_slug"] = custom_slug
        if expires_at is not None:
            body["expires_at"] = expires_at
        if allow_duplicate is not None:
            body["allow_duplicate"] = allow_duplicate
        req_url = self._url("/_/api/links")
        return Link.from_dict(
            self._request("POST", req_url, headers=self._json_headers(), json=body)
        )

    def update(
        self,
        id: int,
        *,
        url: str | None = None,
        label: str | None = None,
        expires_at: int | None = None,
    ) -> Link:
        """Update a link's URL, label, or expiry."""
        body: dict[str, Any] = {}
        if url is not None:
            body["url"] = url
        if label is not None:
            body["label"] = label
        if expires_at is not None:
            body["expires_at"] = expires_at
        req_url = self._url(f"/_/api/links/{id}")
        return Link.from_dict(
            self._request("PUT", req_url, headers=self._json_headers(), json=body)
        )

    def disable(self, id: int) -> Link:
        """Disable a link (stops redirecting)."""
        url = self._url(f"/_/api/links/{id}/disable")
        return Link.from_dict(self._request("POST", url, headers=self._headers()))

    def enable(self, id: int) -> Link:
        """Re-enable a disabled link."""
        url = self._url(f"/_/api/links/{id}/enable")
        return Link.from_dict(self._request("POST", url, headers=self._headers()))

    def delete(self, id: int) -> DeletedResult:
        """Permanently delete a link."""
        url = self._url(f"/_/api/links/{id}")
        return DeletedResult.from_dict(self._request("DELETE", url, headers=self._headers()))

    def analytics(self, id: int, *, range: TimelineRange | None = None) -> ClickStats:
        """Get click analytics for a link."""
        url = self._url(f"/_/api/links/{id}/analytics", {"range": range})
        return ClickStats.from_dict(self._request("GET", url, headers=self._headers()))

    def timeline(self, id: int, *, range: TimelineRange | None = None) -> TimelineData:
        """Get click timeline for a link."""
        url = self._url(f"/_/api/links/{id}/timeline", {"range": range})
        return TimelineData.from_dict(self._request("GET", url, headers=self._headers()))

    def qr(self, id: int, *, slug: str | None = None, size: str | None = None) -> str:
        """Get the QR code SVG for a link. Returns the SVG string."""
        query: dict[str, str | None] = {}
        if slug is not None:
            query["slug"] = slug
        if size is not None:
            query["size"] = size
        url = self._url(f"/_/api/links/{id}/qr", query or None)
        return self._request_text("GET", url, headers=self._headers())

    def bundles(self, id: int) -> _list[Bundle]:
        """List bundles that contain this link."""
        url = self._url(f"/_/api/links/{id}/bundles")
        data = self._request("GET", url, headers=self._headers())
        return [Bundle.from_dict(x) for x in (data or [])]


class AsyncLinks:
    """Asynchronous Links resource."""

    def __init__(self, base_url: str, api_key: str, http: httpx.AsyncClient) -> None:
        self._base_url = base_url
        self._api_key = api_key
        self._http = http

    def _headers(self) -> dict[str, str]:
        return _build_auth_headers(self._api_key)

    def _json_headers(self) -> dict[str, str]:
        return {**self._headers(), "Content-Type": "application/json"}

    def _url(self, path: str, query: dict[str, str | None] | None = None) -> str:
        return _build_url(self._base_url, path, query)

    async def _request(self, method: str, url: str, **kwargs: Any) -> Any:
        try:
            response = await self._http.request(method, url, **kwargs)
        except httpx.RequestError as exc:
            raise ShrtnrError(0, str(exc)) from exc
        return parse_json_response(response)

    async def _request_text(self, method: str, url: str, **kwargs: Any) -> str:
        try:
            response = await self._http.request(method, url, **kwargs)
        except httpx.RequestError as exc:
            raise ShrtnrError(0, str(exc)) from exc
        return parse_text_response(response)

    async def get(self, id: int, *, range: TimelineRange | None = None) -> Link:
        """Get a link by ID. Optional range scopes the click-count window."""
        url = self._url(f"/_/api/links/{id}", {"range": range})
        return Link.from_dict(await self._request("GET", url, headers=self._headers()))

    async def list(
        self, *, owner: str | None = None, range: TimelineRange | None = None
    ) -> list[Link]:
        """List all links. Filter by owner or click-count range."""
        url = self._url("/_/api/links", {"owner": owner, "range": range})
        data = await self._request("GET", url, headers=self._headers())
        return [Link.from_dict(x) for x in (data or [])]

    async def create(
        self,
        *,
        url: str,
        label: str | None = None,
        slug_length: int | None = None,
        custom_slug: str | None = None,
        expires_at: int | None = None,
        allow_duplicate: bool | None = None,
    ) -> Link:
        """Create a new short link."""
        body: dict[str, Any] = {"url": url}
        if label is not None:
            body["label"] = label
        if slug_length is not None:
            body["slug_length"] = slug_length
        if custom_slug is not None:
            body["custom_slug"] = custom_slug
        if expires_at is not None:
            body["expires_at"] = expires_at
        if allow_duplicate is not None:
            body["allow_duplicate"] = allow_duplicate
        req_url = self._url("/_/api/links")
        return Link.from_dict(
            await self._request("POST", req_url, headers=self._json_headers(), json=body)
        )

    async def update(
        self,
        id: int,
        *,
        url: str | None = None,
        label: str | None = None,
        expires_at: int | None = None,
    ) -> Link:
        """Update a link's URL, label, or expiry."""
        body: dict[str, Any] = {}
        if url is not None:
            body["url"] = url
        if label is not None:
            body["label"] = label
        if expires_at is not None:
            body["expires_at"] = expires_at
        req_url = self._url(f"/_/api/links/{id}")
        return Link.from_dict(
            await self._request("PUT", req_url, headers=self._json_headers(), json=body)
        )

    async def disable(self, id: int) -> Link:
        """Disable a link (stops redirecting)."""
        url = self._url(f"/_/api/links/{id}/disable")
        return Link.from_dict(await self._request("POST", url, headers=self._headers()))

    async def enable(self, id: int) -> Link:
        """Re-enable a disabled link."""
        url = self._url(f"/_/api/links/{id}/enable")
        return Link.from_dict(await self._request("POST", url, headers=self._headers()))

    async def delete(self, id: int) -> DeletedResult:
        """Permanently delete a link."""
        url = self._url(f"/_/api/links/{id}")
        return DeletedResult.from_dict(await self._request("DELETE", url, headers=self._headers()))

    async def analytics(self, id: int, *, range: TimelineRange | None = None) -> ClickStats:
        """Get click analytics for a link."""
        url = self._url(f"/_/api/links/{id}/analytics", {"range": range})
        return ClickStats.from_dict(await self._request("GET", url, headers=self._headers()))

    async def timeline(self, id: int, *, range: TimelineRange | None = None) -> TimelineData:
        """Get click timeline for a link."""
        url = self._url(f"/_/api/links/{id}/timeline", {"range": range})
        return TimelineData.from_dict(await self._request("GET", url, headers=self._headers()))

    async def qr(self, id: int, *, slug: str | None = None, size: str | None = None) -> str:
        """Get the QR code SVG for a link. Returns the SVG string."""
        query: dict[str, str | None] = {}
        if slug is not None:
            query["slug"] = slug
        if size is not None:
            query["size"] = size
        url = self._url(f"/_/api/links/{id}/qr", query or None)
        return await self._request_text("GET", url, headers=self._headers())

    async def bundles(self, id: int) -> _list[Bundle]:
        """List bundles that contain this link."""
        url = self._url(f"/_/api/links/{id}/bundles")
        data = await self._request("GET", url, headers=self._headers())
        return [Bundle.from_dict(x) for x in (data or [])]
